src/Models/HydrostaticFreeSurfaceModels/matrix_implicit_free_surface_solver.jl

```diff
@@ -0,0 +1,127 @@
+
+    compute_vertically_integrated_lateral_areas!(vertically_integrated_lateral_areas, grid, arch)
+
+    right_hand_side = arch_array(arch, zeros(eltype(grid), grid.Nx * grid.Ny)) # linearized RHS for matrix operations
+
+    # Set maximum iterations to Nx * Ny if not set
+    settings = Dict{Symbol, Any}(settings)
+    maximum_iterations = get(settings, :maximum_iterations, grid.Nx * grid.Ny)
+    settings[:maximum_iterations] = maximum_iterations
+
+    coeffs = compute_matrix_coefficients(vertically_integrated_lateral_areas, grid, gravity)
+
+    solver = MatrixIterativeSolver(coeffs;
+                            reduced_dim = (false, false, true),
+                                   grid = grid,
+                                   settings...)
+
+    return MatrixImplicitFreeSurfaceSolver(vertically_integrated_lateral_areas, solver, right_hand_side)
+end
+
+build_implicit_step_solver(::Val{:MatrixIterativeSolver}, arch, grid, gravity, settings) =
+    MatrixImplicitFreeSurfaceSolver(arch, grid, gravity, settings)
+
+#####
+##### Solve...
+#####
+
+function solve!(η, implicit_free_surface_solver::MatrixImplicitFreeSurfaceSolver, rhs, g, Δt)
+
+    solver = implicit_free_surface_solver.matrix_iterative_solver
+
+    solve!(η, solver, rhs, Δt)
+
+    return nothing
+end
+
+function compute_implicit_free_surface_right_hand_side!(rhs,
+                                                        implicit_solver::MatrixImplicitFreeSurfaceSolver,
+                                                        g, Δt, ∫ᶻQ, η)
+
+    solver = implicit_solver.matrix_iterative_solver
+    arch = solver.architecture
+    grid = solver.grid
+
+    event = launch!(arch, grid, :xy,
+                    implicit_linearized_free_surface_right_hand_side!,
+                    rhs, grid, g, Δt, ∫ᶻQ, η,
+		            dependencies = device_event(arch))
+
+    return event
+end
+
+# linearized right hand side
+@kernel function implicit_linearized_free_surface_right_hand_side!(rhs, grid, g, Δt, ∫ᶻQ, η)
+    i, j = @index(Global, NTuple)
+    Az   = Azᶜᶜᵃ(i, j, 1, grid)
+    δ_Q  = flux_div_xyᶜᶜᵃ(i, j, 1, grid, ∫ᶻQ.u, ∫ᶻQ.v)
+    t = i + grid.Nx * (j - 1)
+    @inbounds rhs[t] = (δ_Q - Az * η[i, j, 1] / Δt) / (g * Δt)
+end
+
+function compute_matrix_coefficients(vertically_integrated_areas, grid, gravity)
+
+    arch = grid.architecture
+
+    Nx, Ny = (grid.Nx, grid.Ny)
+
+    C     = zeros(Nx, Ny, 1)
+    diag  = arch_array(arch, zeros(eltype(grid), Nx, Ny, 1))
+    Ax    = arch_array(arch, zeros(eltype(grid), Nx, Ny, 1))
+    Ay    = arch_array(arch, zeros(eltype(grid), Nx, Ny, 1))
+    Az    = zeros(Nx, Ny, 1)
+
+    ∫Ax = vertically_integrated_areas.xᶠᶜᶜ
+    ∫Ay = vertically_integrated_areas.yᶜᶠᶜ
+
+    event_c = launch!(arch, grid, :xy, _compute_coefficients!,
+                      diag, Ax, Ay, ∫Ax, ∫Ay, grid, gravity,
+                      dependencies = device_event(arch))
+  
+    wait(event_c)
+
+    return (Array(Ax), Array(Ay), Array(Az), Array(C), diag)
```

Comment:
It is there because the matrix is initially constructed on the CPU so Ax, Ay, Az and C have to be supplied on CPU. Maybe we can change it that we supply the coefficients as we want and move them to the CPU only in the matrix constructor